pandas/core/util/_dispatching.py

```diff
@@ -0,0 +1,186 @@
+
+from pandas._libs import groupby as libgroupby
+from pandas.core.dtypes.missing import isnull
+
+
+class CythonDispatcher(object):
+
+    def __init__(self, groupby):
+        self.groupby = groupby
+        self.func_nm = None
+        self.obj = None
+
+    @property
+    def func_metadata(self):
+        """
+        Stores the metadata required to dispatch each function.
+
+        The format of the dict is as follows:
+
+        attr_name : {
+            'application': {'aggregate', 'transform'}
+            'cython_nm': ...  # Name of the Cython function to call
+            'extra_kwargs': {...}  # Extra kwargs to pass to Cython
+            'type_blacklist': [...]  # Dtypes for which func should raise
+            'result_type': ...  # dtype of result from Cython
+            'conversion_in': ...  # dtype or callable for conversion pre-Cython
+            'needs_values': ...  # Whether the obj values should pass to Cython
+            'needs_mask': ...   # Whether a mask of NA values should be passed
+            'conversion_out': ...  # dtype or callable for conv post-Cython
+        }
+        """
+        return {
+            'any': {
+                'application': 'aggregate',
+                'cython_nm' : 'group_any_all',
+                'extra_kwargs': {'val_test': 'any'},
+                'type_blacklist': [],
+                'result_type': np.uint8,
+                'conversion_in': self._any_all_convertor,
+                'needs_values': True,
+                'needs_mask': True,
+                'conversion_out': np.bool
+            },
+            'all': {
+                'application': 'aggregate',
+                'cython_nm' : 'group_any_all',
+                'extra_kwargs': {'val_test': 'all'},
+                'type_blacklist': [],
+                'result_type': np.uint8,
+                'conversion_in': self._any_all_convertor,
+                'needs_values': True,
+                'needs_mask': True,
+                'conversion_out': np.bool
+            }
+        }
+
+    @property
+    def application_type(self):
+        return self.func_metadata[self.func_nm]['application']
+
+    def _any_all_convertor(self, vals):
+        """
+        Converts objects to appropriate type for any/all calculations.
+        """
+        try:
+            vals = vals.astype(np.bool)
+        except ValueError:  # for objects
+            vals = np.array([bool(x) for x in vals])
+
+        return vals.view(np.uint8)
+
+    def _validate_types(self):
+        """
+        Validate that the types of the `grp_by` object.
+
+        Raises
+        ------
+        ``TypeError`` if the `grp_by` dtypes are not valid for `func_nm`.
+        """
+        if self.obj.values.dtype in self.func_metadata[
+                self.func_nm]['type_blacklist']:
+            raise TypeError("'{}' cannot be applied to a dtype of {}".format(
+                self.func_nm, self.obj.values.dtype))
+
+    def _get_result(self, **kwargs):
```

Comment:
This func does all the heavy lifting; mostly matches the implementation of `_get_cythonized_result` currently in `groupby.py`